android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMKeyboardDownloaderActivity.java

```diff
@@ -113,267 +118,280 @@ public void onCreate(Bundle savedInstanceState) {
           }
+        } else {
+          // Keyman cloud
+          remoteUrl = String.format("%slanguages/%s/%s?device=%s", kKeymanApiBaseURL, langID, kbID, deviceType);
+        }
 
-          kbData = jsonParser.getJSONObjectFromUrl(remoteUrl);
+        ret = downloadNonKMPKeyboard(remoteUrl);
+      } catch (Exception e) {
+        ret = -1;
+        Log.e("Keyboard download", "Error: " + e);
+      }
 
-          exceptionStr = "Could not reach server";
-          if (kbData == null) {
-            throw new Exception(exceptionStr);
-          }
+      return ret;
+    }
 
-          exceptionStr = "The keyboard could not be installed";
-          JSONObject options = kbData.optJSONObject(KMKey_Options);
-          if (options == null) {
-            throw new Exception(exceptionStr);
-          }
-          String kbBaseUri = options.optString(KMKey_KeyboardBaseURI, "");
-          if (kbBaseUri.isEmpty()) {
-            throw new Exception(exceptionStr);
-          }
+    @Override
+    protected void onProgressUpdate(Integer... progress) {
+      // Do nothing
+    }
 
-          String fontBaseUri = options.optString(KMKey_FontBaseURI, "");
-          JSONObject language, keyboard;
-          if (!isCustom) {
-            // JSON from Keyman cloud server
-            language = kbData.optJSONObject(KMKey_Language);
-            if (language == null) {
-              throw new Exception(exceptionStr);
-            }
+    @Override
+    protected void onPostExecute(Integer result) {
+      try {
+        if (progressDialog != null && progressDialog.isShowing()) {
+          progressDialog.dismiss();
+          progressDialog = null;
+        }
+      } catch (Exception e) {
+        progressDialog = null;
+      }
+
+      ((Activity) context).finish();
+      notifyListeners(KeyboardEventHandler.EventType.KEYBOARD_DOWNLOAD_FINISHED, result);
+    }
 
-            langName = language.optString(KMManager.KMKey_Name, "");
+    /**
+     * Download a non-KMP Keyman keyboard.
+     * This will either be from Keyman cloud or legacy ad-hoc distribution via JSON
+     * @param remoteUrl String
+     * @return ret int -1 for fail
+     * @throws Exception
+     */
+    protected int downloadNonKMPKeyboard(String remoteUrl) throws Exception {
+      int ret = -1;
+      JSONParser jsonParser = new JSONParser();
+      JSONObject kbData = jsonParser.getJSONObjectFromUrl(remoteUrl);
+      String exceptionStr = "Could not reach server";
+      if (kbData == null) {
+        throw new Exception(exceptionStr);
+      }
 
-            JSONArray keyboards = language.getJSONArray(KMKey_LanguageKeyboards);
-            if (keyboards == null ) {
-              throw new Exception(exceptionStr);
-            }
+      exceptionStr = "The keyboard could not be installed";
+      JSONObject options = kbData.optJSONObject(KMKey_Options);
+      if (options == null) {
+        throw new Exception(exceptionStr);
```

Comment:
I'm not keen on the single message which doesn't clarify what is wrong with the data ...

Reply:
I'll change that to
```
exceptionStr = "JSON file does not contain a valid \"options\" object";
```